### PR TITLE
[18.0][MIG] product_packaging_level: Migration to 18.0

### PR DESCRIPTION
**Pull request by trisdoan**
### Changes in 18.0

1. get became private since [here](https://github.com/odoo/odoo/pull/118701/commits/2b93736d5cd2a8bcba9cb612a4e680b6cd7f3854) 
2. _lang_get_xxx  were merged into one api _get_data which is simpler and easy to extend since [here](https://github.com/odoo/odoo/commit/9f5470f247526fe54ff86e931cc835ef8942132a) 
3. Odoo 17.0+: `product_id` is required on model product_packaging since [here](https://github.com/odoo/odoo/commit/0e71e350e35744c38a8c70c2d8c75b39f5be0262)
As consequences,
   - no need default value in [_compute_qty_per_level](https://github.com/trisdoan/product-attribute/pull/2/commits/998cd1fffbb91a6a809b89f7fac57faadeeb9d45)
   - Some little changes on test setup
      - test packagings are assigned with a same product

### CODE REVIEW

File: product_packaging_level/views/product_packaging_view.xml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        <field name="model">product.packaging</field>
+        <field name="inherit_id" ref="stock.product_packaging_tree_view" />
+        <field name="arch" type="xml">
+            <field name="name" position="before">
+                <field name="display_name" string="Packaging" />
+            </field>
+            <field name="package_type_id" position="after">
+                <field name="qty_per_level" />
+                <field name="packaging_level_id" />
+                <field name="barcode_required_for_gtin" invisible="1" />
```
Comment (nguyenminhchien) [inline product_packaging_level/views/product_packaging_view.xml]:
```suggestion
                <field name="barcode_required_for_gtin" column_invisible="1" />
```

File: product_packaging_level/views/product_packaging_view.xml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            </field>
+            <xpath expr="//field[@name='barcode']" position="attributes">
+                <attribute name="required">barcode_required_for_gtin</attribute>
+            </xpath>
+            <field name="name" position="attributes">
+                <attribute name="invisible">1</attribute>
```
Comment (nguyenminhchien) [inline product_packaging_level/views/product_packaging_view.xml]:
```suggestion
                <attribute name="column_invisible">1</attribute>
```

File: product_packaging_level/models/product_packaging.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                # We have to allow several packaging using the default type,
+                # because when we install the module on an existing database,
+                # the default value will be set to default and we'll have
+                # duplicates. Anyway "default" is not meant to be used as a
+                # real type.
+                if not packaging.packaging_level_id.is_default
+            ]
+            if len(set(packaging_level_ids)) != len(packaging_level_ids):
+                raise ValidationError(
+                    _(
```
Comment (nguyenminhchien) [inline product_packaging_level/models/product_packaging.py]:
```suggestion
                    self.env._(
```

File: product_packaging_level/models/product_packaging_level.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        ),
+    )
+
+    @api.constrains("name_policy")
+    def _check_packaging_name(self):
+        for packaging in self:
+            activated_packages = self.env.user.has_group("stock.group_tracking_lot")
+            if packaging.name_policy == "by_package_type" and not activated_packages:
+                raise ValidationError(
+                    _(
```
Comment (nguyenminhchien) [inline product_packaging_level/models/product_packaging_level.py]:
```suggestion
                    self.env._(
```

File: product_packaging_level/models/product_packaging_level.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                        " Configuration > Settings !"
+                    )
+                )
+
+    @api.constrains("is_default")
+    def _check_is_default(self):
+        msg = False
+        default_count = self.search_count([("is_default", "=", True)])
+        if default_count == 0:
+            msg = _('There must be one product packaging level set as "Is Default".')
```
Comment (nguyenminhchien) [inline product_packaging_level/models/product_packaging_level.py]:
```suggestion
            msg = self.env._('There must be one product packaging level set as "Is Default".')
```

File: product_packaging_level/models/product_packaging_level.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        elif default_count > 1:
+            msg = _('Only one product packaging level can be set as "Is Default".')
```
Comment (nguyenminhchien) [inline product_packaging_level/models/product_packaging_level.py]:
```suggestion
            msg = self.env._('Only one product packaging level can be set as "Is Default".')
```

**Review by nguyenminhchien — CHANGES_REQUESTED** [review]
You should start from this PR: https://github.com/OCA/product-attribute/pull/1662

**Comment by nguyenminhchien** [comment]
> You should start from this PR: [OCA#1662](https://github.com/OCA/product-attribute/pull/1662)

You could also review this PR

**Comment by nguyenminhchien** [comment]
This commit https://github.com/trisdoan/product-attribute/commit/18b6d8a1433d9b1424cf3be73616d4e14c96dd77 breaks the user-define rule for naming of Packaging: User cannot input name
image

File: product_packaging_level/views/product_packaging_view.xml (hunk 1 above)
Comment (trisdoan) [reply product_packaging_level/views/product_packaging_view.xml]: Hi anh, I decided to remove it as:

> Missing fields used in domains and other attributes are added to the view as invisible automatically

**Comment by trisdoan** [comment]
> This commit [18b6d8a](https://github.com/trisdoan/product-attribute/commit/18b6d8a1433d9b1424cf3be73616d4e14c96dd77) breaks the user-define rule for naming of Packaging: User cannot input name image

Hi anh, indeed, it blocks user to input name even with user_define policy
I created an extra commit (easily to pick to former versions), dont you agree?

**Review by nguyenminhchien — APPROVED** [review]

File: product_packaging_level/models/product_packaging_level.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+class ProductPackagingLevel(models.Model):
+    _name = "product.packaging.level"
+    _description = "Level management for product.packaging"
+    _order = "sequence, code"
+
+    def _default_language(self):
+        lang_code = self.env["ir.default"]._get("res.partner", "lang")
+        def_lang_id = self.env["res.lang"]._get_data(code=lang_code).id
+        return def_lang_id or self._active_languages()[0]
```
Comment (thienvh332) [inline product_packaging_level/models/product_packaging_level.py:15]: `_active_languages` ​​does not exist in `product.packaging.level`.

Comment (trisdoan) [reply product_packaging_level/models/product_packaging_level.py:15]: thanks, it's fixed

**Pull request closed without merging**